streaming/base/dataset.py

```diff
@@ -765,6 +787,128 @@ def _resample_streams(self, epoch: int) -> Tuple[NDArray[np.int64], NDArray[np.i
+
+        batch_size = self.batch_size or 1
+
+        for stream_id, stream in enumerate(self.streams):
+            shuffle_units, small_per_big = self._resample_streams(epoch, stream_id)
+            samples_in_stream = len(small_per_big)
+            stream_partition = get_partitions(self.partition_algo, samples_in_stream,
+                                              self.num_canonical_nodes, world.num_nodes,
+                                              world.ranks_per_node, world.workers_per_rank,
+                                              batch_size, 0)
+            if self.shuffle:
+                # Ratio of stream's shuffle block size to overall shuffle block size should be the
+                # same as the ratio of the stream's samples to overall samples.
+                # This ensures that the overall training shuffle block size is still approximately
+                # equal to what is set by the user, and allows for reasoning about cache_limit as well.
+                shuffle_block_portion = int(self.shuffle_block_size * stream.proportion)
+                stream_shuffle = get_shuffle(self.shuffle_algo, shuffle_units,
+                                             self.num_canonical_nodes, self.shuffle_seed, epoch,
+                                             shuffle_block_portion)
+                stream_partition = np.where(stream_partition != -1,
+                                            stream_shuffle[stream_partition], -1)
+            # The small_per_big array already corresponds to indices of samples per shard of each stream.
+            # So each sample ID in the stream's partition already corresponds to the sample ID in the right shard.
+            partition_per_stream.append(
+                np.where(stream_partition != -1, small_per_big[stream_partition], -1))
+
+        # We now merge the partitions from each stream to get our final partition over all streams, where
+        # each global batch has samples only from a single stream.
+        # Partitions are arranged (physical nodes, ranks per node, workers per rank, batches per worker, batch size).
+        batches_per_stream = []
+        batches_from_partitions = []
+        for i, partition in enumerate(partition_per_stream):
+            # reshape the partition to be global batches in order of traversal, and count only batches without -1 in them.
+            global_batches_inorder = partition.transpose(3, 2, 0, 1, 4).reshape(
+                -1, batch_size * world.ranks_per_node * world.num_nodes)
+            num_full_batches = np.count_nonzero(np.min(global_batches_inorder, axis=1) >= 0)
+            batches_per_stream.append(num_full_batches)
+            if num_full_batches != global_batches_inorder.shape[0]:
+                logger.warning(
+                    'Because of the `per_stream` batching method, some batches with an inadequate number of samples '
+                    + 'from stream with index ' + str(i) + ' are being dropped.')
```

Comment:
is `str(i)` a sample index ? I am trying to understand, if this error message gets displays with `str(i)` information, how much can user infer what it is saying !!

Reply:
`str(i)` is the index of the stream's partition in the partition_per_stream array. i'll change the variable name to reflect that